index.ts

```diff
@@ -41,7 +41,9 @@ type AnalyticsEventType = 'identify' | 'track' | 'page' | 'screen' | 'group' | '
 export default class Analytics {
   private readonly enable: boolean;
 
-  private readonly queue = [] as {
+  private flushInProgress: boolean = false;
+  private readonly flushQueue: (() => Promise<FlushResponse>)[] = [];
```

Comment:
Rather than a flush queue and needing to see if the message queue length is divisible by flushAt, I think it will be simpler to write the logic like:

```
fn flush() {
  get a batch of messages (determined by both flushAt and maxFlushSizeInBytes)
  send batch to server
  if another batch is ready or there's no timer (indicating that the timer expired and is waiting for the in-flight flush to finish)
    flush()
```

The flush queue goes away and `flush()` runs until it's caught up to a normal state. The tricky part is figuring out what the behavior of calling `flush()` should be while there are in-flight flushes. Should it finish when all in-flight flushes are done? Should it do one extra flush after they're done? What is the behavior of concurrent flush() calls?

Reply:
i'll take a look tomorrow and see if this will work.

Reply:
If we were to adopt your approach above - the behavior i could see working is 
```
Promise<FlushResponse> currentFlush;

...

fn flush() {
  are we already flushg?
    return await currentFlush
  currentFlush = await executeFlush()
  return currentFlush
}

fn executeFlush(){
get a batch of messages (determined by both flushAt and maxFlushSizeInBytes)
  send batch to server
  add this batch of messages to the accumulated flush response
  if another batch is ready or there's no timer & items in the queue (indicating that the timer expired and is waiting for the in-flight flush to finish)
    executeFlush - pass in partial results up to this point
  else call flush callback, return accumulated flush response
}
```
^^ flush would need a return type of Promise<FlushResponse[]> afaict

> The tricky part is figuring out what the behavior of calling flush() should be while there are in-flight flushes. Should it finish when all in-flight flushes are done? 

I believe it would need to `finish` when all flushes have been executed. If it 'finishes' after only the first batch is sent, we could enter the scenario where there are items left 'unflushed' and the caller exits the process thinking all items had been sent. (e.g. `exit() { await rudderClient.flush(); process.exit(); }` only sends the first batch of messages, not all messages)

> Should it do one extra flush after they're done? 

If flush is reworked so it calls itself until there are no messages left in the queue, i don't see a reason for this. I'm assuming our check to flush again inside of the flush will catch all cases we would want to flush for.

If, on the other hand, we continue with the bound flush approach, then we would match the number of flush calls to the number of flushes executed (if no messages are in the queue, we'd just exit early).

> What is the behavior of concurrent flush() calls?

If there is no longer a flush queue, i believe the only behavior we could have would be 
- the first flush to start will handle sending all the messages in the queue
- later flush calls will have to `return await` a reference to the results of the first (recursive) flush 

to lower the complexity we could also ditch callbacks for `flush()`. I think now that the caller can await, there isn't any real need. We would still provide the callback interface for track/identify/etc...

Reply:
OK. that sounds reasonable.

I am still hesitant about prescheduling flushes based on multiples of flushAt. For instance, if flushAt = 20 and 40 messages are enqueued, two flushes will be scheduled. If the first flush fails, we'll retry again (with the new splice-on-success logic). In any case, we're left with messages in the queue after the second flush attempt.

Reply:
yup. seems there's 2 questions here --
- whether a 'flush till empty' algorithm will work better for us than a 'flush as directed' algorithm
- how to deal with messages that fail to send 

-- And both options for question #1 are affected by how to deal with message failures. 

So, focusing on the behavior we want when handling message failures:

I think it's a likely scenario for network connection failures to cause batches to fail to send. whether the situation resolves itself or not would depend on the user's context. For example, if someone is developing an expo app on a plane with airplane mode on, I would not want our library to constantly be sending HTTP requests. 

I also think it's not unlikely that our self-hosted solution goes down from time to time. For example, if we need to update the Postgres instance to use a larger capacity, there will be ~3-5 mins of downtime as google requisitions a new instance. In these cases, it would be good to store the events for as long as we can.

IMO - the most important thing to do is detect when we are in an infi-loop and halt it
e.g. we don't want `await flush()` to stop `eas-cli` from exiting. 

We can avoid getting into a loop in the first place by dropping events that fail to send (simplest solution). But assuming we retain them, we'll just need to eat the complexity of handling infinite loops (catches some events that would have otherwise been dropped).

Some things we could do
- drop events if they fail to send (under any circumstances)
- drop events if they fail to send due to network connection (still retry on 4xx/5xx) -- might cause the CLI to hang on occasion
- infinite loop detection
  - we could just say `maxFlushAttempts = items in queue / batchSize ...` and exit early if we need to
  - or add some logic to check # of times a message/batch has failed to flush, drop it at a certain point

maybe a happy medium is- make retries take longer/have more attempts, drop the messages after they fail.

IDK, any pref for you?

As per the flush algorithm, at this point, i think either both will work. The main impetus behind binding the pending flushes was to 
- preserve a mapping of specific flush => specific callback (which it doesn't seem like actually care about)
- preserve our flush signature of `Promise<FlushResponse>`, which is equivalent to a single response of the HTTP batch API (which we also don't really need to care about?)
But for sure it's a bit more complex and memory intensive.

Reply:
I like this holistic approach to figure out the behavior we want. I agree we want to be careful with the number of retries. In addition to the scenarios you laid out, there could also be a malformed message that always gets rejected and needs to be dropped.

For the scenarios with transient errors (network failure, server outage), exponential backoff is a go-to approach. For permanent or not-so-transient errors, we might just want to drop the messages if we get a server error or pause if we get a network error. Or drop in both cases since it probably doesn't matter.

For flushing I think it makes a lot of sense for the behavior to be: when you call flush(), the returned promise will be fulfilled when all of the events currently in the queue (but not ones that might be enqueued afterward) have been handled (sent/dropped).

Reply:
> For the scenarios with transient errors (network failure, server outage), exponential backoff is a go-to approach. For permanent or not-so-transient errors, we might just want to drop the messages if we get a server error or pause if we get a network error. Or drop in both cases since it probably doesn't matter.

I have a feeling that it's not possible for us to know when an error is transient or not transient. For example, losing network connection could be a hiccup, or it could be someone who is developing in airplane mode. Same for a server outage, which might be 2 minutes or 2 days (knock on wood). As long as it sounds reasonable to you, I think i'd like to go ahead with processing each message once and dropping it whether it's a failure or success.

I think the only exception to the above is if we get a `429` - rate limited. But we already have some # of retries, so I think we can just bump the retry threshold a bit and call it a day. Nice and simple.

Reply:
> For flushing I think it makes a lot of sense for the behavior to be: when you call flush(), the returned promise will be fulfilled when all of the events currently in the queue (but not ones that might be enqueued afterward) have been handled (sent/dropped).

We can definitely accomplish this, but it does add a decent amount of complexity to the flush logic itself. Consider:
- a flush is already in flight
- the user adds a new message 
- the user calls `flush()`

we cannot just return a reference to the pending flush (since it will explicitly not process that new message). We would need to either: enqueue a flush (similar to what is currently in this PR) or modify the in-flight flush to also process that message. I think the latter is the simplest solution and will move forward with that as my approach.

Reply:
Retrying a bit with exponential backoff (which we already do) is probably good enough. If we can't deliver a message after making a good effort, it seems fine for our needs to drop it.

For flushing: in this scenario we would wait for the in-flight flush to complete and then flush another batch that includes the message the developer intends to flush when they called `flush()`.